src/librustc_data_structures/smallvec.rs

```diff
@@ -71,25 +71,26 @@ impl<T> FromIterator<T> for MaybeSmallVec8<T> {
     fn from_iter<I: IntoIterator<Item=T>>(iter: I) -> MaybeSmallVec8<T> {
         let iter = iter.into_iter();
         if iter.size_hint().1.map_or(false, |n| n <= 8) {
-            let mut count = 0;
-            let mut values = [
-                ManuallyDrop { empty: () },
-                ManuallyDrop { empty: () },
-                ManuallyDrop { empty: () },
-                ManuallyDrop { empty: () },
-                ManuallyDrop { empty: () },
-                ManuallyDrop { empty: () },
-                ManuallyDrop { empty: () },
-                ManuallyDrop { empty: () },
-            ];
+            let mut v = SmallVec8 {
+                count: 0,
+                values: [
+                    ManuallyDrop { empty: () },
+                    ManuallyDrop { empty: () },
+                    ManuallyDrop { empty: () },
+                    ManuallyDrop { empty: () },
+                    ManuallyDrop { empty: () },
+                    ManuallyDrop { empty: () },
+                    ManuallyDrop { empty: () },
+                    ManuallyDrop { empty: () },
+                ],
+            };
             for el in iter {
-                values[count] = ManuallyDrop { value: el };
-                count += 1;
+                unsafe {
+                    v.values[v.count].value = el;
+                }
+                v.count += 1;
```

Comment:
If you make this loop an `Extend` impl, you can then have `fill_item` take `T: Extend<Kind> + Deref<Target=[Kind]>`.